java/sage/plugin/CorePluginManager.java

```diff
@@ -2315,6 +2192,268 @@ public static File getDevPluginsDir() {
+      if (localExtract.isFile() && !canModifyPath(localExtract))
+      {
+        openFailed = true;
+      }
+      String orgMD5 = null;
+      if (Sage.DBG) System.out.println("Current extraction local target=" + localExtract);
+      if (localExtract.isFile())
+      {
+        if (pendingFilesystem.containsKey(localExtract))
+        {
+          orgMD5 = (String) pendingFilesystem.get(localExtract);
+          if ("0".equals(orgMD5))
+          {
+            orgMD5 = null;
+          } else
+          {
+            trackFile(localExtract, priorStateIndex, orgMD5, true);
+          }
+        } else
+        {
+          orgMD5 = IOUtils.calcMD5(localExtract);
+          // Update the filetracker with the apriori state of this file
+          // This will also store a backup of the file if needed
+          trackFile(localExtract, priorStateIndex, orgMD5, true);
+        }
+      } else
+      {
+        java.io.File dirParent = localExtract.getParentFile();
+        if (!dirParent.isDirectory() && !dirParent.mkdirs())
+        {
+          if (Sage.DBG) System.out.println("ERROR Cannot create directory from zip archive: " + dirParent);
+          progressMsg = "";
+          return "FAILED - Cannot Create " + dirParent;
+        }
+        if (Sage.DBG)
+          System.out.println("Ensured existence of parent directory " + dirParent + " isDir=" + dirParent.isDirectory());
+        // Also track any parent directories that were automatically created
+        String dirName = entry.getName();
+        int dirIdx = dirName.lastIndexOf(java.io.File.separatorChar);
+        while (dirIdx != -1)
+        {
+          dirName = dirName.substring(0, dirIdx);
+          trackFile(new java.io.File(destDir, dirName), myInstallIndex, "0", false);
+          dirIdx = dirName.lastIndexOf(java.io.File.separatorChar);
+        }
+      }
+
+      FileOutputStream outStream=null;
+      if (!openFailed)
+      {
+        try
+        {
+          outStream = new java.io.FileOutputStream(localExtract);
+        } catch (java.io.IOException ioe)
+        {
+          if (localExtract.isFile())
+          {
+            if (Sage.DBG) System.out.println("ERROR opening local file to write to: " + localExtract);
+            // Extract the file to a temporary location and check the MD5Sum of it to see if we even need to install it
+            openFailed = true;
+          } else
+            throw ioe;
+        }
+      }
+      if (openFailed)
+      {
+        localExtract = new java.io.File(localExtract.getAbsolutePath() + "." + myInstallIndex);
+        int x = 0;
+        while (localExtract.isFile())
+        {
+          localExtract = new java.io.File(localExtract.getAbsolutePath() + "." + myInstallIndex + "." + x++);
+        }
+        outStream = new java.io.FileOutputStream(localExtract);
+      }
+      InputStream inStream = entry.getInputStream();
```

Comment:
This no longer gets closed if an exception occurs, we had a finally block for that before